kernel/src/process.rs

```diff
@@ -84,6 +69,130 @@ pub unsafe fn load_processes(
     }
 }
 
+/// This trait is implemented by process structs.
+pub trait ProcessType {
+    /// Queue a callback for the process. This will be added to a per-process
+    /// buffer and passed to the process by the scheduler.
+    fn schedule(&self, callback: FunctionCall) -> bool;
+
+    /// Queue an IPC operation for this process.
+    fn schedule_ipc(&self, from: AppId, cb_type: IPCType);
+
+    /// Remove the scheduled operation from the front of the queue.
+    fn unschedule(&self) -> Option<Task>;
+
+    /// Returns the current state of the process. Common states are "running" or
+    /// "yielded".
+    fn current_state(&self) -> State;
+
+    /// Move this process from the running state to the yield state.
+    fn yield_state(&self);
+
+    /// Put this process in the fault state. This will trigger the
+    /// `FaultResponse` for this process to occur.
+    unsafe fn fault_state(&self);
```

Comment:
bikeshed: The verbage across `current_state`, `yield_state` and `fault_state` is inconsistent, the first is a `get` while the second two are `set` operations. I'd prefer the interface make this explicit: `get_current_state`, and ideally `set_state(&self, state: State)`, but that doesn't work if the fault needs to be unsafe and the yield doesn't..., why is `fault_state` and unsafe function? Is that actually Rust unsafe, or a we-dont-have-capabilities-merged-yet unsafe?

Reply:
I think the it _should_ be the latter. logically this is just changing a field (whether actual or logical) in a type that has interior mutability, so from a Rust safety perspective the implementation should be safe.

In the current implementation (in master) `fault_state` is doing double duty---it's switching `&self` into the faulted state, and it's also clearing the global flag that the `arch` crate is using to communicate to the `kernel` crate that there was _a_ fault from _a_ process

if we cleanly separate those two pieces of functionality, I think we can just have `state` and `set_state` methods.

Reply:
There is an unsafe operation in `fault_state` when trying to restart a faulted process: it needs to calculate where the init_fn is using the start of flash, the init fn offset from the TBF header, and the `offset` function. The `offset` function is unsafe.

Reply:
Right, but the important question is whether it exposes a safe interface, not whether it uses unsafe under the hood.

Reply:
In any case I tried merging `yield_state` and `fault_state`, but then the API includes the ability to set the state as `Running`, which isn't a valid use of the API. So, I think it is better to leave them as separate functions, just renamed (`set_yielded_state`, `set_fault_state`).